metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -532,113 +722,105 @@ func join(fields ...string) string {
+		watcher.watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				update(enricher.metadata, obj.(kubernetes.Resource))
+			},
+			UpdateFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				update(enricher.metadata, obj.(kubernetes.Resource))
+			},
+			DeleteFunc: func(obj interface{}) {
+				enricher.Lock()
+				defer enricher.Unlock()
+				delete(enricher.metadata, obj.(kubernetes.Resource))
+			},
+		})
+	}
 
 	return &enricher
 }
 
-func (m *enricher) Start() {
-	m.watchersStartedLock.Lock()
-	defer m.watchersStartedLock.Unlock()
-	if !m.watchersStarted {
-		if m.nodeWatcher != nil {
-			if err := m.nodeWatcher.Start(); err != nil {
-				logp.Warn("Error starting node watcher: %s", err)
-			}
-		}
-
-		if m.namespaceWatcher != nil {
-			if err := m.namespaceWatcher.Start(); err != nil {
-				logp.Warn("Error starting namespace watcher: %s", err)
-			}
-		}
+func (e *enricher) Start(resourceWatchers *Watchers) {
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
 
-		if m.replicasetWatcher != nil {
-			if err := m.replicasetWatcher.Start(); err != nil {
-				logp.Warn("Error starting replicaset watcher: %s", err)
-			}
+	resourceWatcher := resourceWatchers.watchersMap[e.resourceName]
+	if resourceWatcher != nil && resourceWatcher.watcher != nil && !resourceWatcher.started {
+		if err := resourceWatcher.watcher.Start(); err != nil {
+			log.Warnf("Error starting %s watcher: %s", e.resourceName, err)
+		} else {
+			resourceWatcher.started = true
 		}
+	}
 
-		if m.jobWatcher != nil {
-			if err := m.jobWatcher.Start(); err != nil {
-				logp.Warn("Error starting job watcher: %s", err)
+	extras := getExtraWatchers(e.resourceName, e.config)
+	for _, extra := range extras {
+		extraWatcher := resourceWatchers.watchersMap[extra]
+		if extraWatcher != nil && extraWatcher.watcher != nil && !extraWatcher.started {
+			if err := extraWatcher.watcher.Start(); err != nil {
+				log.Warnf("Error starting %s watcher: %s", extra, err)
+			} else {
+				extraWatcher.started = true
 			}
 		}
-
-		err := m.watcher.Start()
-		if err != nil {
-			logp.Warn("Error starting Kubernetes watcher: %s", err)
-		}
-		m.watchersStarted = true
 	}
 }
 
-func (m *enricher) Stop() {
-	m.watchersStartedLock.Lock()
-	defer m.watchersStartedLock.Unlock()
-	if m.watchersStarted {
-		m.watcher.Stop()
-
-		if m.namespaceWatcher != nil {
-			m.namespaceWatcher.Stop()
-		}
-
-		if m.nodeWatcher != nil {
-			m.nodeWatcher.Stop()
-		}
+func (e *enricher) Stop(resourceWatchers *Watchers) {
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
 
-		if m.replicasetWatcher != nil {
-			m.replicasetWatcher.Stop()
+	resourceWatcher := resourceWatchers.watchersMap[e.resourceName]
+	if resourceWatcher != nil && resourceWatcher.watcher != nil && resourceWatcher.whichAreUsing != nil && resourceWatcher.started {
```

Comment:
How can `resourceWatcher.whichAreUsing` be nil if watcher is started ?

Reply:
You are right, if `resourceWatcher` is not nil, then the data inside (watcher, resourcesUsing and started) should exist too. I will delete those conditions